documents/development/events.md

```diff
@@ -0,0 +1,108 @@
+
+The backend is responsible for:
+
+- connecting to the data source
+- managing the creation of it’s own tables
+- implementing new classes which extend from `EventPrototype` and `EventBackendPrototype` and implementing the required methods, ie `save` and `static findByGuid`. See `core/api/classes/events` to learn more.
+
+## Schema
+
+The 3 models used for the Grouparoo Event System are: `events`, `eventData` and `apiKeys`. While the specific implementing of the schema used by each backend may differ, a prototypical table structure is modeled below.
+
+### `events` table
+
+This is the primary event table logging a row per event.
+
+- _guid_ - The primary Key for this event. Generated by the Grouparoo server. Should /not/ be auto-incremented but others unique (we use UUIDv4) + a `evt_` prefix.
+- _producerGuid_ - “foreign key” for `apiKeys`. Identifies the API client which created the event, ie: “web API client” vs “iPhone API client”. Set by the grouparoo server.
+- _ipAddress_ - IP address the event was received from. Set by the grouparoo server.
+- _profileGuid_ - “foreign key” for `profiles`. Set by the grouparoo server.
+- _createdAt_ - The time the Grouparoo Server received the event. Set by the grouparoo server.
+- _updatedAt_ - The time the Grouparoo Server last modified the event. Set by the grouparoo server.
+- _occurredAt_ - Set by the client, optional. This likely differs from `createdAt` as it should store the local time of the event. Event’s may be batched when sent from the client, so we cannot rely on the time server received the event to be the time the event occurred. If there is no `occurredAt`, the server will apply `createdAt`
+- _type_ - Set by the client, required. The named type of the event. Examples include `pageview`, `item-added-to-cart`, `purchase`, or `click`
+- _userId_ - Set by the client, optional. If the client knows the primary identifying profile property of the current user, it should be sent. This is /your/ notion of userId. This property takes precedence over `anonymousId` when present.
+- _anonymousId_ - Set by the client, required (unless `userId` is present). If the `userId` is not known (the user is not yet logged in), a unique ID should be generated by the client. These can be device GUIDs in the case of mobile phones, or just another UUID.
+
+### `eventData` table
+
+This is a “has-many” data table logging the data properties of each event. Each event can have 0 or many rows in this table.
+
+- _guid_ - Set by the Grouparoo server. The primary Key for this portion of an event’s data. Generated by the Grouparoo server. Should /not/ be auto-incremented but others unique (we use UUIDv4) + a `edt_` prefix.
+- _key_- Set by the client, required. The named key of this portion of an event’s data.
+- _value_- Set by the client, required. The value of this portion of an event’s data. All data is stored as strings.
```

Comment:
want to `rawValue` here like in profile properties?

Reply:
I don't think so.  The reason it works with Profile Properties is we know how to cast back because of the Profile Property Rule.  We'd need something like that for each possible eventData#key to cast property.  I think I'd rather move the casting step to the source, and cast based on the typeof the profile property rule you are trying to make

Reply:
I'm not sure of the answer here. I agree, but it would be nice to know the type in the source somehow to make profile property creation easier. Probably not worth it so they just have to pick the type when making the rule as you said.